### PR TITLE
support sdk types

### PR DESCRIPTION
**Pull request by carmelc**
Currently the library assumes that the types are under `types` entry in `package.json` and that the library only has a single export entry (that there is no `import {context} from 'lib-name/context';` for example,
The purpose of the PR is to support more options:
1. several package.json files under the same library
2. `exports` definition in `package.json`
3. The package.json scheme does not define exactly how types are exposed - `typings`/`types` so the PR adds support for both
4. Introduce a new package.json definition - `typesBundle` which defines the bundled definition so there is no need to run the whole bundler again, this was added for all metro auto generated sdk packages.

### CODE REVIEW

File: src/cache.ts [hunk 1]
```diff
@@ -26,16 +26,16 @@ export function cacheFactoryFactory<Key, Value>() {
...
-    
+
         return cache;
     }
 
     return {
         cacheFactory,
         async clearCache() {
-            await Promise.all(Array.from(factoryCaches.values()).map(cache => cache.clear()));
+            // await Promise.all(Array.from(factoryCaches.values()).map(cache => cache.clear()));
```
Comment (yurynix) [inline src/cache.ts]: Forgot to uncomment?

File: src/index.ts [hunk 2]
```diff
@@ -21,35 +25,35 @@ export async function bundle(packageIdentifier: string, outputFilePath: string,
         let resultCode = undefined;
         await tempy.directory.task(async tempDirectory => {
             const saveFileFromPackage = createFetcher(cache.cacheFactory);
-            const pkgJsonData: any = await saveFileFromPackage(tempDirectory, packageName, packageVersion, 'package.json');
-            const pkgJson = JSON.parse(pkgJsonData);
+            const loadFileForPackage = (filePath: string, content?: string) => saveFileFromPackage(tempDirectory, packageName, packageVersion, filePath, content);
+            const typesEntryContent = await createPackageTypes(packageName, packageVersion, loadFileForPackage);
+            // load the generated file to the cache so it is
```
Comment (yurynix) [inline src/index.ts]: so it is...? what? 🙃

Comment (carmelc) [reply src/index.ts]: this is what you get when NOT using GPT...

File: __tests__/e2e/simple-e2e.spec.ts [hunk 3]
```diff
@@ -8,73 +8,127 @@ describe('Types bundler', () => {
...
+            export = yurypkg;
+        }
+        `));
     });
 
     it('should work with external package dependency with relative .. path import', async () => {
         const mockPackages = {
             ...pkg('yury-pkg@1.0.0', `import { Foo } from 'foo-pkg';
-            export type Bla { one: Foo };`, { 'foo-pkg': '1.2.3' }),
+            export type Bla { one: Foo };`, {dependencies: { 'foo-pkg': '1.2.3' }}),
```
Comment (yurynix) [inline __tests__/e2e/simple-e2e.spec.ts:37]: `pkg` implementation is:
```ts
export const pkg = (pkgIdentifier: string, typesContents: string, {
  dependencies = {},
  typesName = 'types'
} = {}) => {
...
  const pkgJson = {
    [`${pkgIdentifier}/package.json`]: JSON.stringify({
      name: packageName,
      version: packageVersion,
      [typesName]: types,
      dependencies: {
        ...dependencies
      },
    })
```

so why `dependencies: ...` and not like it was, without the key `dependnecies`?

File: __tests__/e2e/wix-sdk-e2e.latest-versions.spec.ts [hunk 4]
```diff
@@ -0,0 +1,92 @@
+import tmp from 'tmp';
+import { bundle } from '../../src';
+import fs from 'fs';
+import { getLatestVersionName, validateTypescript } from './utils';
+
+describe('Latest versions', () => {
```
Comment (yurynix) [inline __tests__/e2e/wix-sdk-e2e.latest-versions.spec.ts]: Why this is called `Latests versions` ? 🤔 
The tests validate asterisks exports, etc.

File: src/fetch-unpkg.ts [hunk 5]
```diff
@@ -56,34 +38,37 @@ function createDataFetcher(cacheFactory: CacheFactory<string, Promise<string>>)
...
-        const overallCacheKey = `${url}|${filePath}`;
-        if (await overallCache.has(overallCacheKey)) {
-            debug(`Using cached ${overallCacheKey}`);
-            const cacheResult = await overallCache.get(overallCacheKey);
-            return cacheResult as string;
+    return async function saveFileFromPackage(rootDir, packageName, packageVersion, filePath, fileContent) : Promise<string> {
+        // remove leading '/'
+        const adjustedFilePath = filePath.replace(rootDir, '').replace(/^\/+/, '');
+        if (/^(\.ts|\.d\.ts|.*\/\.ts|.*\/\.d\.ts)$/.test(adjustedFilePath)) {
+            return Promise.reject('');
```
Comment (yurynix) [inline src/fetch-unpkg.ts]: Maybe something like `.reject(new Error('Unable to adjust file path'))`; ?

or `Adjusted filepath not valid file for bundling` or someting

**Review by yurynix — APPROVED** [review]
Thank you for working on this,
LGTM.

Had a few minor comments for your consideration.

File: __tests__/e2e/simple-e2e.spec.ts (hunk 3 above)
Comment (carmelc) [reply __tests__/e2e/simple-e2e.spec.ts:37]: the dependencies are in an object

**Pull request merged**